#163 Architecture design for transaction history (#63)
digitalaidseattle/plymouth-housing: docs/designs/transaction-design.md

            Transactions th
        INNER JOIN 
            Items i ON th.item_id = i.item_id
        ORDER BY 
            th.transaction_date DESC;      -- Optionally order by transaction date or any other field
    END
    ELSE
    BEGIN
        -- If not an admin, return an error message
        THROW 50000, 'Access Denied. Only admins can access transaction history.', 1;
    END
END;
```

### API Endpoints

**Record Transaction**
- **Endpoint:** POST /api/transactions
- **Request Body:**  
  ```json
  {
    "userID": 123,
    "transactionGUID": "abc123",
    "totalAmount": 100.00,
    "transactionType": "Add",
    "transactionDate": "2021-10-01T12:00:00",
    "itemID": 456
  }
  ```
- **Response:**  
  ```json
  {
    "transactionID": 789
  }
  ```
  
**Get Transaction History**
- **Endpoint:** GET /api/transactions?userID=123
- **Response:**  
  ```json
  [
    {
      "transactionID": 789,
      "userID": 123,
      "transactionGUID": "abc123",
      "totalAmount": 100.00,
      "transactionType": "Add",
      "transactionDate": "2021-10-01T12:00:00",
      "itemID": 456
    }
  ]
  ```

### Data Flow
#### When a transaction occurs (items added or checked out):

> - Front-end sends a POST request to the `/api/transactions` endpoint with transaction details and generates a TransactionGUID for the transaction.

Reviewer: Suggestion to change this to a stored procedure so that only one api call is required to change the inventory and log a transaction at the same time.

- API endpoints for CRUD transaction data.
- A front-end page for displaying transaction history to admin users.

## Architecture Goals and Constraints

### Goals
1. Ensure transaction history is recorded with:
    - Primary key for each transaction.
    - User ID.
    - Transaction GUID.
    - Item ID.
    - Transaction type (e.g., Add, Checkout).
    - Total amount.
    - Transaction date.
2. Provide API endpoints for recording and retrieving transaction data only for admin users.

### Constraints
- Admin role management will control access to the transaction history data.
- Transaction data must be logged in real-time during any item addition or checkout.
- The transaction history page should be able to display all transactions for a specific user.
- The transaction history page should be able to display detailed information for each transaction.
- The transaction history page should be able to filter and search transactions based on various criteria.

## Detailed Design  

### Database Schema

#### Database  

**Tables:**  
- **TransactionHistory**:  
  - `TransactionID` (Primary Key) - Unique identifier for each transaction record (UUID or GUID).

>   - `UserID` (Foreign Key) - References the `Users` table to identify the admin user associated with the transaction.

Reviewer 2: ```suggestion
  - `UserID` (Foreign Key) - References the `Users` table to identify the user (admin or volunteer) associated with the transaction.
```

Reviewer 2: This leads to a good observation. We don't have the capability to track the admin email for transactions. We should. [#196](https://das-ph-inventory-tracker.atlassian.net/browse/PIT-196)

>   - `TransactionGUID`- Unique identifier for each transaction session (UUID or GUID). Used to correlate multiple entries for the same transaction (e.g., for the same user interacting multiple times within one session).

Reviewer 2: ```suggestion
  - `TransactionGUID`- Unique identifier for each transaction session (UUID or GUID). Used to correlate multiple entries for the same transaction (e.g., for the same user checking out several items within one transaction, or tracking all the items in a welcome basket).
```

  - `ItemID` (Foreign Key): References the `Items` table to identify the item associated with the transaction.
  - `TransactionType`: VARCHAR(50), Type of transaction (e.g., Add, Checkout).

>   - `TotalAmount`: DECIMAL(10, 2), Total amount of the transaction.

Reviewer 2: ```suggestion
  - `Quantity`: DECIMAL(10, 2), Total number of items with *ItemId* in this transaction.
```

  - `TransactionDate`: DATETIME, Date and time when the transaction occurred.

**SQL Schema:**
```sql
CREATE TABLE TransactionHistory (

>     TransactionID UNIQUEIDENTIFIER PRIMARY KEY,       

Reviewer 2: ```suggestion
        id INT NOT NULL IDENTITY(1,1) PRIMARY KEY,,       
```
This is what we use in the other tables. So for consistency...

>     UserID INT,                                        

Reviewer 2: ```suggestion
    UserID INT NOT NULL,                                        
```

>     TransactionGUID UNIQUEIDENTIFIER,                  

Reviewer 2: ```suggestion
    TransactionID UNIQUEIDENTIFIER  NOT NULL,                  
```

>     ItemID INT,                                         

Reviewer 2: ```suggestion
    ItemID INT  NOT NULL,                                         
```

>     TransactionType VARCHAR(50),                          

Reviewer 2: ```suggestion
    TransactionType VARCHAR(50)  NOT NULL,                          
```

>     TotalAmount DECIMAL(10, 2),                         

Reviewer 2: ```suggestion
    Quantity INT  NOT NULL,                         
```
I think this will always be an integer. We could verify though

>     TransactionDate DATETIME DEFAULT GETDATE(),         

Reviewer 2: ```suggestion
    TransactionDate DATETIME DEFAULT GETDATE()  NOT NULL,         
```

    
    -- Define foreign key constraints
    FOREIGN KEY (UserID) REFERENCES Users(UserID),      

>     FOREIGN KEY (ItemID) REFERENCES Items(ItemID)       

Reviewer 2: Excellent. We should do this in other tables too!

);

-- Optionally, we may want to create an index for faster searching by TransactionGUID, UserID, and TransactionDate
CREATE INDEX IX_TransactionHistory_TransactionGUID ON TransactionHistory(TransactionGUID);
CREATE INDEX IX_TransactionHistory_UserID ON TransactionHistory(UserID);

> CREATE INDEX IX_TransactionHistory_TransactionDate ON TransactionHistory(TransactionDate);

Reviewer 2: Great idea. Let's wait till we hit performance issues though. But this is the one table that will get very long quickly.

```

**Stored Procedures:**
- **RecordTransaction**:  
  - Inserts a new transaction record into the `TransactionHistory` table.

>   - Parameters: `UserID`, `TransactionGUID`, `ItemID`, `TransactionType`, `TotalAmount`.

Reviewer 2: Can you match the parameter names with my suggestions throughout the document (if you agree with the names)

  
```sql
CREATE PROCEDURE LogTransaction
    @UserID INT,                     
    @TransactionGUID UNIQUEIDENTIFIER, 
    @ItemID INT,                   
    @TransactionType VARCHAR(50),     
    @TotalAmount DECIMAL(10, 2)       
AS
BEGIN
    -- Insert a new record into the TransactionHistory table
    INSERT INTO TransactionHistory (
        TransactionID,
        UserID,
        TransactionGUID,
        ItemID,
        TransactionType,
        TotalAmount,
        TransactionDate
    )
    VALUES (
        NEWID(),                     -- Generate a new GUID for the TransactionID
        @UserID,                     
        @TransactionGUID,           
        @ItemID,                   
        @TransactionType,             
        @TotalAmount,               
        GETDATE()                 
    );
END;
```

> - **GetTransactionHistory**:

Reviewer 2: This is a great suggestion. I would like to hold off on implementing these until we have the user stories and the figma designs for the transaction history
- do we want them by user, building, date, category, etc.
- are we going to do client side filtering or server side.

- Retrieves all transactions associated with a specific `UserID`(admin user only).

```sql
CREATE PROCEDURE GetTransactionHistory

>     @AdminUserID INT                 -- The ID of the admin requesting transaction history

Reviewer 2: ```suggestion
    @UserID INT                 -- The ID of the admin requesting transaction history
```
I think we want them for volunteers as well

AS
BEGIN
    -- Check if the requesting user is an admin
    IF EXISTS (SELECT 1 FROM Users WHERE UserID = @AdminUserID AND IsAdmin = 1)
    BEGIN
        -- Retrieve all transaction history records
        SELECT 
            th.TransactionID,
            th.UserID,
            th.TransactionGUID,
            th.ItemID,
            th.TransactionType,
            th.TotalAmount,
            th.TransactionDate
        FROM 
            TransactionHistory th
        INNER JOIN 
            Items i ON th.ItemID = i.ItemID
        ORDER BY 
            th.TransactionDate DESC;      -- Optionally order by transaction date or any other field
    END
    ELSE
    BEGIN
        -- If not an admin, return an error message
        THROW 50000, 'Access Denied. Only admins can access transaction history.', 1;
    END
END;
```

> ### API Endpoints

Reviewer 2: These are easily exposed through the json file with the other REST API endpoints.
